Stop using dispatchTypeId to do checks for tensor list unwrap.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33011 Beef up documentation on DispatchKey.h
* #32788 Add XLAPreAutograd key for XLA use cases that need custom autograd.
* **#32787 Stop using dispatchTypeId to do checks for tensor list unwrap.**

Gets rid of a longstanding TODO.  TensorList unwrap is only used for cat, which
means we can assume that the inputs are dense, and do something similar to how
we do the dense tensor wrapping above.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628642](https://our.internmc.facebook.com/intern/diff/D19628642)

ezyang committed:
Stop using dispatchTypeId to do checks for tensor list unwrap.

Gets rid of a longstanding TODO.  TensorList unwrap is only used for cat, which
means we can assume that the inputs are dense, and do something similar to how
we do the dense tensor wrapping above.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit f81e7519:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f81e75195a8b0ac7efb29caa53641e96ee4d43cd).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4372428) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92771794) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92771794))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
HEAD is now at f81e75195a Stop using dispatchTypeId to do checks for tensor list unwrap. 
+ git merge --allow-unrelated-histories --no-edit --no-ff 8cb05e72c69fdd837548419770f3f1ba9807c16d 
Auto-merging c10/core/DispatchKey.h 
CONFLICT (content): Merge conflict in c10/core/DispatchKey.h 
Removing aten/src/THNN/generic/BCECriterion.c 
Removing aten/src/THCUNN/generic/BCECriterion.cu 
Removing aten/src/THCUNN/BCECriterion.cu 
Auto-merging aten/src/ATen/function_wrapper.py 
Removing aten/src/ATen/detail/ScalarTypeConversions.h 
Auto-merging aten/src/ATen/core/dispatch/DispatchKeyExtractor.h 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang committed:
Update on "Stop using dispatchTypeId to do checks for tensor list unwrap."


Gets rid of a longstanding TODO.  TensorList unwrap is only used for cat, which
means we can assume that the inputs are dense, and do something similar to how
we do the dense tensor wrapping above.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop using dispatchTypeId to do checks for tensor list unwrap."


Gets rid of a longstanding TODO.  TensorList unwrap is only used for cat, which
means we can assume that the inputs are dense, and do something similar to how
we do the dense tensor wrapping above.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628642](https://our.internmc.facebook.com/intern/diff/D19628642)

[ghstack-poisoned]

ezyang committed:
Update on "Stop using dispatchTypeId to do checks for tensor list unwrap."


Gets rid of a longstanding TODO.  TensorList unwrap is only used for cat, which
means we can assume that the inputs are dense, and do something similar to how
we do the dense tensor wrapping above.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628642](https://our.internmc.facebook.com/intern/diff/D19628642)

[ghstack-poisoned]

ezyang committed:
Update on "Stop using dispatchTypeId to do checks for tensor list unwrap."


Gets rid of a longstanding TODO.  TensorList unwrap is only used for cat, which
means we can assume that the inputs are dense, and do something similar to how
we do the dense tensor wrapping above.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628642](https://our.internmc.facebook.com/intern/diff/D19628642)

[ghstack-poisoned]
